### PR TITLE
Add E2E test suite for cephfs

### PR DESCRIPTION
**Pull request by Madhu-1**
Added E2e test framework

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>
Co-authored-by: Poornima G pgurusid@redhat.com

### CODE REVIEW

**Comment by Madhu-1** [comment]
@poornimag PTAL

**Comment by Madhu-1** [comment]
planning to use rook [framework](https://github.com/rook/rook/blob/master/tests/integration/base_deploy_test.go#L103) to deploy rook and to create cephfs and rbd pools

need to do code refractor in (https://github.com/rook/rook/blob/master/tests/integration) ( move some functions to a new go file) in rook to import rook test framework in ceph-csi

@travisn is this acceptable in rook?

**Comment by Madhu-1** [comment]
Finally, being able to run the test for cephfs and rbd
```
Ran 2 of 2 Specs in 487.358 seconds
SUCCESS! -- 2 Passed | 0 Failed | 0 Pending | 0 Skipped
--- PASS: TestE2E (487.37s)
PASS
ok  	github.com/ceph/ceph-csi/e2e	487.491s
```

**Comment by Madhu-1** [comment]
[csi-test.log](https://github.com/ceph/ceph-csi/files/3186219/csi-test.log) Attaching logs for more info

**Comment by Madhu-1** [comment]
https://github.com/rook/rook/pull/3192

File: .travis.yml [hunk 1]
```diff
@@ -50,11 +51,15 @@ jobs:
...
       script:
         - make cephcsi ||  travis_terminate 1;
+        - make image-cephcsi || travis_terminate 1;
         - sudo scripts/minikube.sh up || travis_terminate 1;
         # pull docker images to speed up e2e
         - scripts/minikube.sh cephcsi
         - scripts/minikube.sh k8s-sidecar
         - "sudo chown -R travis: $HOME/.minikube /usr/local/bin/kubectl"
+        # functional tests
+        - make func-test TESTOPTIONS='--rook-version=v1.0.0 \
```
Comment (humblec) [inline .travis.yml]: `v1.1.0` ?

File: Gopkg.toml [hunk 2]
```diff
@@ -14,9 +14,21 @@
...
+  name = "k8s.io/cli-runtime"
+
+[[override]]
+  version = "kubernetes-1.14.0"
+  name = "k8s.io/apiextensions-apiserver"
+
 [[constraint]]
   name = "k8s.io/kubernetes"
-  version = "v1.14.0"
+  version = "=v1.14.0"
```
Comment (humblec) [inline Gopkg.toml]: I have to read through. whats this `=` does here? limiting to exact version?

File: e2e/cephfs.go [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+import (
+	"time"
+
+	. "github.com/onsi/ginkgo" // nolint
+
+	"k8s.io/kubernetes/test/e2e/framework"
+)
+
+var (
+	cephProvisioner     = "csi-cephfsplugin-provisioner.yaml"
```
Comment (humblec) [inline e2e/cephfs.go]: cephfsProvisioner.. and same for below vars.

File: e2e/cephfs.go [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+}
+
+func deployCephFSNodePlugin() {
+	//TODO need to update the image name
+	framework.RunKubectlOrDie("create", "-f", cephfsDirPath+cephNodePlugin)
+	//create provisoner RBAC
+	framework.RunKubectlOrDie("apply", "-f", cephfsDirPath+cephNodePluginRBAC)
+}
+
```
Comment (humblec) [inline e2e/cephfs.go]: may be make deployment together in a same function?

File: e2e/rbd.go [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+}
+
+func deployRBDNodePlugin() {
+	//TODO need to update the image name
+	framework.RunKubectlOrDie("create", "-f", rbdDirPath+rbdNodePlugin)
+	//create provisoner RBAC
+	framework.RunKubectlOrDie("apply", "-f", rbdDirPath+rbdNodePluginRBAC)
+}
+
```
Comment (humblec) [inline e2e/rbd.go]: Same as above

**Comment by humblec** [comment]
@Madhu-1 I would suggest squash commits of code in one and vendor updates in another commit, that will make it easy to review .

File: .travis.yml (hunk 1 above)
Comment (Madhu-1) [reply .travis.yml]: This is the current stable version in rook, so using stable release

File: Gopkg.toml (hunk 2 above)
Comment (Madhu-1) [reply Gopkg.toml]: yes, it will lock  to exact  version, I have faced a lot of issues (i thought  fixing to exact  version is a better option)

**Comment by Madhu-1** [comment]
@humblec @ShyamsundarR PTAL

**Comment by Madhu-1** [comment]
i will add  a readme and remove DNM  label later

File: .travis.yml (hunk 1 above)
Comment (travisn) [reply .travis.yml]: `v1.0.1` is the latest stable as of today

Comment (Madhu-1) [reply .travis.yml]: @travisn Thanks, updated to `v1.0.1`

File: e2e/e2e_test.go [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+//BeforeSuite deploys the rook-operator and ceph cluster
+var _ = BeforeSuite(func() {
+	if rookRequired {
+		deployRook()
+	}
+})
+
+//AfterSuite removes the rook-operator and ceph cluster
+var _ = AfterSuite(func() {
```
[the remark this answers is not in the document]
Comment (Madhu-1) [reply e2e/e2e_test.go]: will this help in Travis? making teardown optional  may leave  some undeleted  resource
if required I think we can update it later.

**Comment by Madhu-1** [comment]
Added readme @humblec @ShyamsundarR PTAL

**Comment by humblec** [comment]
@Madhu-1 Thanks for adding README and adding the test suite. Lets have this as the starting point and do rest or improve it later.!

**Pull request merged**